File: cpp/3_group_anagrams.cpp
/**
 * LeetCode #49: 字母异位词分组
 * 题目: 给定一个字符串数组，将字母异位词组合在一起。
 * 解法: 哈希表 - 将每个字符串排序后作为 key，字母异位词排序后相同，
 *       因此可以加入同一组。
 * 时间 O(n * klogk)，空间 O(n*k)。k 为字符串平均长度。
 */

#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> groups;
        for (const string& s : strs) {
            string key = s;
            sort(key.begin(), key.end());
            groups[key].push_back(s);
        }
        vector<vector<string>> result;
        for (auto& kv : groups) {
            result.push_back(kv.second);
        }
        return result;
    }
};

int main() {
    Solution sol;
    vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
    auto result = sol.groupAnagrams(strs);
    for (const auto& group : result) {
        cout << "[";
        for (int i = 0; i < group.size(); ++i) {
            cout << group[i];
            if (i < group.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
    }
    return 0;
}

File: python/3_group_anagrams.py
#!/usr/bin/env python3
"""
LeetCode #49: 字母异位词分组
题目: 给定一个字符串数组，将字母异位词组合在一起。
解法: 哈希表 - 将每个字符串排序后作为 key，字母异位词排序后相同，
      因此可以加入同一组。时间 O(n * klogk)，空间 O(n*k)。
      k 为字符串平均长度。
"""


from collections import defaultdict


def group_anagrams(strs: list[str]) -> list[list[str]]:
    """
    排序法: 将字符串排序得到规范形式作为字典 key。
    """
    groups = defaultdict(list)
    for s in strs:
        key = ''.join(sorted(s))
        groups[key].append(s)
    return list(groups.values())


if __name__ == "__main__":
    # 测试用例
    print(group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))
    # [['eat', 'tea', 'ate'], ['tan', 'nat'], ['bat']]
